### ISSUE TITLE
updatePickerInput will not work with delays/long processing time in server

### ISSUE THREAD

**Issue by s5joschi** [rstudio/shiny#3004]
I'm trying to create inputs that are updated based on other input fields with the `updatePickerInput` function. The code works as expected if the user e.g. deselects slowly inputs from `vs` in the example below. However, if the user is very quick in deselecting, the app will start jumping forth and back between different inputs. This problem increases, if processing time is increased in the ` server `, which I do artificially in the example below with `sleep.time`. I really cannot figure out what the problem is and if this is more general problem that might affect other applications as well. Any help would be much appreciated. 

```R
library(shiny)
library(dplyr)
library(plotly)
library(shinyWidgets)
library(shinydashboard)

data("mtcars")

# create ui
ui <- fluidPage(fluidRow(
  box(
    title = "Filter",
    # define the picker inputs
    pickerInput(
      inputId = "cyl_pickerinput",
      label = "CYL",
      choices = levels(as.factor(mtcars$cyl)),
      multiple = T,
      selected = levels(as.factor(mtcars$cyl)),
    ),
    pickerInput(
      inputId = "vs_pickerinput",
      label = "VS",
      choices = levels(as.factor(mtcars$vs)),
      multiple = T,
      selected = levels(as.factor(mtcars$vs))
    ),
    pickerInput(
      inputId = "carb_pickerinput",
      label = "CARB",
      choices = levels(as.factor(mtcars$carb)),
      multiple = T,
      selected = levels(as.factor(mtcars$carb))
    )
  ),
  box(title = "Plot of mtcars",
      plotlyOutput("carplot"))
))

# create server
server <- function(input, output, session) {
  #(1) Create PickerInput Updates
  observeEvent(
    # define pickerinputs to be observed
    c(
      input$vs_pickerinput,
      input$carb_pickerinput,
      input$cyl_pickerinput
    ),
    {
      ## update PickerInput based on a condition that requires the user to choose at least one input
      if (!is.null(input$cyl_pickerinput) &
          !is.null(input$vs_pickerinput) &
          !is.null(input$carb_pickerinput)) {
        # if there is no null input filter values
        mtcars.filter <-
          mtcars %>%
          filter(cyl %in% input$cyl_pickerinput) %>% # filters
          filter(vs %in% input$vs_pickerinput) %>%
          filter(carb %in% input$carb_pickerinput)
        # update the pickers 
        updatePickerInput(
          session = session,
          inputId ="cyl_pickerinput",
          choices = levels(factor(mtcars$cyl)),
          selected = mtcars.filter$cyl)
        updatePickerInput(
          session = session,
          inputId ="carb_pickerinput",
          choices = levels(factor(mtcars$carb)),
          selected = mtcars.filter$carb)
        updatePickerInput(
          session = session,
          inputId ="vs_pickerinput",
          choices = levels(factor(mtcars$vs)),
          selected  = mtcars.filter$vs)
        # render the plot
        output$carplot <-
          renderPlotly({
            # create plot
            plot <- ggplot(mtcars.filter) +
              geom_point(aes(wt, mpg, color = factor(vs)))
            # convert to plotly
            ggplotly(plot)
          })
        # or alternatively for reproducing the problem: introduce latency with sys.sleep
        Sys.sleep(1)
        
      }
    },ignoreInit = TRUE
  )
}

shinyApp(ui, server)

```

**Comment by cpsievert** [comment]
Your code/description is hard to follow (your code errors out and uses packages not listed here, such as `shinyWidgets`), but it sounds like you may want to `throttle()` or `debounce()` the updating logic?

**Comment by s5joschi** [comment]
@cpsievert. I'm sorry for that. I had actually two libraries missing. Now it should work, I tested it in two different machines. Should I explain in more detail what the code does? 

Also I was searching around the last couple of hours and I guess the problem could be related to [execution scheduling](https://shiny.rstudio.com/articles/execution-scheduling.html)...? 

So in my case above, inputs in the `ui` trigger an update in the `observeEvent` block in `server` which again triggers an update in the `ui`. It's kind of a two-way dependency. 

In the case, that a user is faster in updating the ui before the server does, he triggers a new update in server, which is only executed once the already running process is through (which again passes new information to the input). 

Does that make sense?

**Comment by cpsievert** [comment]
Yea, I see the problem now, you're on the right track. Thanks for fixing and elaborating. 

In this case, I think you might be able to avoid the problem by using a `renderUI()` pattern rather than an `observe()`+`update()` pattern. Hopefully, this'll help de-couple the UI updating logic from the server computation that you need to do as well

```r

ui <- fluidPage(fluidRow(
  box(
    title = "Filter",
    uiOutput("cyl_pickerinput"),
    uiOutput("vs_pickerinput")
  ),
  box(title = "Plot of mtcars",
      plotlyOutput("carplot"))
))

server <- function(input, output, session) {
  mtcars_filtered <- reactive({
    if (!length(input$cyl_pickerinput)) {
      return(mtcars)
    }
    mtcars %>%
      filter(cyl %in% input$cyl_pickerinput) %>%
      filter(vs %in% input$vs_pickerinput)
  })
  output$carplot <- renderPlotly({
    plot <- ggplot(mtcars_filtered()) +
      geom_point(aes(wt, mpg, color = factor(vs)))
    ggplotly(plot)
  })
  output$vs_pickerinput <- renderUI({
    pickerInput(
      inputId = "vs_pickerinput",
      label = "VS",
      choices = unique(mtcars$vs),
      multiple = T,
      selected = unique(mtcars_filtered()$vs)
    )
  })
  output$cyl_pickerinput <- renderUI({
    pickerInput(
      inputId = "cyl_pickerinput",
      label = "CYL",
      choices = unique(mtcars$cyl),
      multiple = T,
      selected = unique(mtcars_filtered()$cyl)
    )
  })
}

shinyApp(ui, server)
```

**Comment by cpsievert** [comment]
(If you need further help, feel free to open a https://community.rstudio.com/ thread, but I don't think this is necessarily a problem that we can address directly in shiny itself)

**Comment by s5joschi** [comment]
Thanks @cpsievert for clarifying and giving some example code to solve. Glad it is nothing that affects other apps as well! Awesome support. Thanks again!
